### PR TITLE
2816: Redesign the TestSetup API

### PR DESCRIPTION
**Pull request by TobyRet**
### Context
Further to the work undertaken here https://github.com/DFE-Digital/teacher-training-api/pull/1097 a discussion with devs followed and it was agreed that the TestSetup API should be changed to mirror, more closely, that of FactoryBot.

### Changes proposed in this pull request
- Simplification of the TestSetup API and test coverage
- Rather predefining variable names for cached factory instances, user now simply call a new instance method as follows:
`get(factory_type, [traits])`

### Guidance to review
This is the second iteration so as always, keen to get feedback. Hopefully the changes are heading the right direction and the API is now easier to use.

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by TobyRet** [comment]
Will clean up the comments once approved

**Comment by TobyRet** [comment]
[[S] Redesign the TestSetup API](https://trello.com/c/Zlnmu1ku/2816-s-redesign-the-testsetup-api)

File: spec/factories/courses.rb [hunk 1]
```diff
@@ -210,7 +210,7 @@
       to_create { |instance| instance.save(validate: false) }
     end
 
-    trait :unpublished_with_primary_maths do
```
Comment (timabell) [inline spec/factories/courses.rb]: 1. Did you check there were no usages of this trait before renaming it?
2. Do we need to rename it? It would be a cleaner patch with the rename.

File: spec/models/provider_spec.rb [hunk 2]
```diff
@@ -268,7 +268,8 @@
   end
 
   describe "courses" do
-    let(:course) { TestSetup.unpub_pri_math }
+    let(:test_setup) { TestSetup.new }
+    let(:course) { test_setup.get(:course, :primary, :unpublished) }
```
Comment (timabell) [inline spec/models/provider_spec.rb]: I think rename `test_setup` to `test_data`

Comment (timabell) [inline spec/models/provider_spec.rb]: :heart: love the new signature :+1:

File: spec/support/test_setup.rb [hunk 3]
```diff
@@ -2,118 +2,54 @@
...
-      @@courses&.key?(method_name) ||
-      super
-  end
-
   attr_reader :courses, :recruitment_cycles
 
-  def initialize(with_courses: nil)
+  def initialize
     @recruitment_cycles = @@recruitment_cycles.transform_values(&:call)
     @courses = @@courses.transform_values(&:call)
```
Comment (timabell) [inline spec/support/test_setup.rb]: I'd think I'd extract a named method for these two lines. `def CreateAndCacheTestRecords`

File: spec/factories/courses.rb [hunk 4]
```diff
@@ -210,7 +210,7 @@
...
+    trait :unpublished do
       transient do
         identifier { "unpublished_with_primary_maths" }
```
Comment (defong) [inline spec/factories/courses.rb]: whats with primary maths?

File: spec/support/test_setup.rb [hunk 5]
```diff
@@ -2,118 +2,54 @@
...
   }
 
   @@courses = {
-    unpub_pri_math: ->(attributes = {}) do
-      find_or_create(:course, :unpublished_with_primary_maths, attributes)
-    end,
+      [:primary, :unpublished] => ->() do
+        FactoryBot.find_or_create(:course, :primary, :unpublished)
+      end
   }
```
Comment (faissalMT) [inline spec/support/test_setup.rb]: Why not make these methods with switches? It doesn't seem like you're passing around the procs here and it might better reflect what you're trying to do.

File: spec/support/test_setup.rb [hunk 6]
```diff
@@ -2,118 +2,54 @@
...
-      @courses[method_name]
+  def get(name, *traits)
+    case name
+    when :course
+      @courses[traits.sort] || raise(
+          <<~ERR_MSG
+        No predefined course for these traits: #{traits}.
+        Either add it to test_setup.rb or if it's used frequently, just create
+        a FactoryBot factory instance.
+      ERR_MSG
```
Comment (timabell) [inline spec/support/test_setup.rb]: I think I'd DRY the `raise`

```ruby
def UnknownTraits(type_name, traits)
    raise(
          <<~ERR_MSG
        No predefined #{type_name} for these traits: #{traits}.
        Either add it to `test_setup.rb` to take advantage of the caching,
        or create a FactoryBot factory instance directly for one-offs.
      ERR_MSG
end
```

Also iterated on the message

File: spec/support/test_setup.rb [hunk 7]
```diff
@@ -2,118 +2,54 @@
 class TestSetup
```
Comment (timabell) [inline spec/support/test_setup.rb]: I'm thinking we should rename this to `TestDataCache`

File: spec/support/test_setup.rb [hunk 8]
```diff
@@ -2,118 +2,54 @@
...
   # A helper class for instantiating a standard set of factories for testing
```
Comment (timabell) [inline spec/support/test_setup.rb]:
```suggestion
  # A helper class for pre-creating test data with FactoryBot factories
  # and maintaining a cache of these objects
  # to avoid the performance problems of
  # repeatedly asking FactoryBot to create trees of objects.
  # Refs: https://evilmartians.com/chronicles/testprof-2-factory-therapy-for-your-ruby-tests-rspec-minitest
```

File: spec/test_setup_specs/test_setup_spec.rb [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+describe 'TestSetupSpec' do
+  before(:all) do
+    @test_setup = TestSetup.new
+  end
+
+  it 'finds trait pre-defined in test setup [:primary, :unpublished]' do
+    course = @test_setup.get(:course, :primary, :unpublished)
+
+    expect(course).to eq(Course.first)
```
Comment (timabell) [inline spec/test_setup_specs/test_setup_spec.rb]: Will this survive as we pre-cache more variations of course?

Comment (timabell) [reply spec/test_setup_specs/test_setup_spec.rb]: I think maybe we should just test that it does return a course and that it has the right properties

File: spec/factories/courses.rb (hunk 1 above)
Comment (TobyRet) [reply spec/factories/courses.rb]: This was one that I created in iteration 1

File: spec/models/provider_spec.rb (hunk 2 above)
Comment (TobyRet) [reply spec/models/provider_spec.rb]: 👍

File: spec/support/test_setup.rb (hunk 3 above)
Comment (TobyRet) [reply spec/support/test_setup.rb]: 🤔 mmm ... I prefer to declare instance vars in the initializer if possible as I think alternative creates more in-direction. Happy to hear opinions on this though

File: spec/support/test_setup.rb (hunk 6 above)
Comment (TobyRet) [reply spec/support/test_setup.rb]: 👍

File: spec/support/test_setup.rb (hunk 7 above)
Comment (TobyRet) [reply spec/support/test_setup.rb]: Yeah that seems reasonable. Or more more simply 'TestData'? What do others think?

File: spec/support/test_setup.rb (hunk 8 above)
Comment (TobyRet) [reply spec/support/test_setup.rb]: 👍

File: spec/test_setup_specs/test_setup_spec.rb (hunk 9 above)
Comment (TobyRet) [reply spec/test_setup_specs/test_setup_spec.rb]: Yeah agree 👍

File: spec/factories/courses.rb (hunk 4 above)
Comment (TobyRet) [reply spec/factories/courses.rb]: Fixing ta!

File: spec/factory_specs/course_spec.rb [hunk 10]
```diff
@@ -31,8 +34,8 @@
     end
   end
 
-  context "unpublished_with_primary_maths" do
-    subject { create(:course, :unpublished_with_primary_maths) }
+  context "unpublished" do
+    subject { create(:course, :unpublished) }
```
[the remark this answers is not in the document]
Comment (TobyRet) [reply spec/factory_specs/course_spec.rb]: The `:unpublished` trait is also creating sites and associating the same provider with the course and sites. 

We still need to agree what test data is made available through the TestDataCache.

File: spec/support/test_setup.rb (hunk 5 above)
Comment (timabell) [reply spec/support/test_setup.rb]: Resolved by renaming the factory collections to make it clearer what they are for :+1:

File: spec/factory_specs/course_spec.rb [hunk 11]
```diff
@@ -5,6 +5,9 @@
 
   it { should be_instance_of(Course) }
   it { should be_valid }
+  it "has the correct number of associations" do
```
[the remark this answers is not in the document]
Comment (timabell) [reply spec/factory_specs/course_spec.rb]: > shouldn't it be more like this?

no because the default course factory doesn't add sites

[the remark this answers is not in the document]
Comment (defong) [reply spec/factory_specs/course_spec.rb]: can we have a test to denote that then, it just look so similar...

File: spec/support/test_data_cache.rb [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+  # It does not matter what order the traits are when passed as arguments.
+
+
+  class << self
+    def get(name, *traits)
+      case name
+      when :course
+        @@courses[traits.sort] || raise_trait_error(:course, traits)
+      when :recruitment_cycle
+        @@recruitment_cycles[traits.sort] || raise_trait_error(:recruitment_cycle, traits)
```
Comment (misaka) [inline spec/support/test_data_cache.rb]: Don't still need this.

File: spec/test_setup_specs/test_data_cache_spec.rb [hunk 13]
```diff
@@ -0,0 +1,46 @@
+require "rspec"
+
+describe "TestSetupSpec" do
```
Comment (misaka) [inline spec/test_setup_specs/test_data_cache_spec.rb]:
```suggestion
describe TestDataCache do
```

File: config/settings/test.yml [hunk 14]
```diff
@@ -3,3 +3,4 @@ govuk_notify:
   welcome_email_template_id: please_change_me
 system_authentication_token: "Ge32"
 gcp_api_key: please_change_me
+test_setup: false
```
Comment (misaka) [inline config/settings/test.yml]: I don't really feel like this is an app concern ... tests should only ever be run in the `testing` Rails env, for example. I think a bespoke env var would be more appropriate, no?

File: spec/factories/courses.rb [hunk 15]
```diff
@@ -210,12 +210,12 @@
...
 
-    trait :unpublished_with_primary_maths do
+    trait :unpublished do
       transient do
-        identifier { "unpublished_with_primary_maths" }
+        identifier { "unpublished" }
       end
 
-      name { "#{identifier} course name" }
+      name { "#{identifier}_course" }
```
Comment (misaka) [inline spec/factories/courses.rb]: This is what I was talking about before, btw, about using the string "name" in here so that we can trace how this value is used through the system. I think what was here before is right:

```suggestion
      name { "#{identifier} course name" }
```

File: spec/support/test_data_cache.rb [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+class TestDataCache
+  #  A helper class for pre-creating test data with FactoryBot factories
+  #  and maintaining a cache of these objects to avoid the performance problems
+  #  of repeatedly asking FactoryBot to create trees of objects.
+  #
+  #  Refs: https://evilmartians.com/chronicles/testprof-2-factory-therapy-for-your-ruby-tests-rspec-minitest
+  #
+  # The test data and can be accessed as follows:
+  #
+  # TestSetupCache.get(:course, :primary, :unpublished)
```
Comment (misaka) [inline spec/support/test_data_cache.rb]:
```suggestion
  # TestDataCache.get(:course, :primary, :unpublished)
```

File: spec/support/test_data_cache.rb [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+  # Note that a course with the requested traits needs to be pre-defined in this class.
+  # It does not matter what order the traits are when passed as arguments.
+
+
+  class << self
+    def get(name, *traits)
+      case name
+      when :course
+        if @@courses == nil
+          # cache not populated, fall back to factory (while we migrate all our tests)
```
Comment (misaka) [inline spec/support/test_data_cache.rb]: Would it make more sense to test this on a factory-by-factory basis? Something like the following, replacing the `if @@courses == nil` check above.

```ruby
          sorted_traits = traits.sort
          unless @@courses.has_key? sorted_traits
            if @@course_factories.has_key? sorted_traits
              @@courses[sorted_traits] = @@course_factories[sorted_traits].call
            else
              raise_trait_error(:course, traits)
            end
          end
          @@courses[sorted_traits]
```

If you agree, it's probably worth putting most of the above in a separate function.

File: config/settings/test.yml (hunk 14 above)
Comment (TobyRet) [reply config/settings/test.yml]: 👍

File: spec/factories/courses.rb (hunk 15 above)
Comment (TobyRet) [reply spec/factories/courses.rb]: 👍

File: spec/support/test_data_cache.rb (hunk 16 above)
Comment (TobyRet) [reply spec/support/test_data_cache.rb]: 👍

File: spec/support/test_data_cache.rb (hunk 17 above)
Comment (TobyRet) [reply spec/support/test_data_cache.rb]: I'm don't think this is necessary ...
Unless the caching is switched on then `@@courses` will always be nil. Therefore it probably makes sense to simply return the factory.
What do you think @timabell?

File: spec/test_setup_specs/test_data_cache_spec.rb (hunk 13 above)
Comment (TobyRet) [reply spec/test_setup_specs/test_data_cache_spec.rb]: 👍

File: spec/support/test_data_cache.rb (hunk 17 above)
Comment (timabell) [reply spec/support/test_data_cache.rb]: What @TobyRet said.

Because the behaviour is:

* The cache can be turned off/on with the env
* If it's on, then the cache is filled in advance so no need to do granular testing
* If it's off, we bypass the cache entirely and fall back to the factories, so again no need to test what is cached

Having said that, this isn't immediately obvious from the code as written, I think maybe we should add a `cachePopulated` field to the class and set that in the setup method and then update this `if` to use it for clarity

Comment (TobyRet) [reply spec/support/test_data_cache.rb]: ... and / or we can check if the cache is actually turned on? `ENV['TEST_DATA_CACHE']`

Comment (timabell) [reply spec/support/test_data_cache.rb]: As I mentioned on the phone, that make the behaviour unnecessarily affected by external state.

I'm having a go at implementing this but my first attempt broke something...

Comment (timabell) [reply spec/support/test_data_cache.rb]: Refactored, all green. Should be clearer now.

**Comment by timabell** [comment]
Just waiting for @misaka 's :+1: now
:pray:

File: spec/support/test_data_cache.rb [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+
+  class << self
+    def get(name, *traits)
+      case name
+      when :course
+        if @@cache_populated
+          @@courses[traits.sort] || raise_trait_error(:course, traits)
+        else
+          # fall back to factory (while we migrate all our tests)
+          FactoryBot.create(name, *traits)
```
Comment (misaka) [inline spec/support/test_data_cache.rb]: We should be used the factory as defined in `@@course_factories` so we have a single source of truth for how to create this cached data.

```suggestion
          @@course_factories[traits.sort].call
```

File: spec/support/test_data_cache.rb [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+        end
+      else
+        raise "Unknown model type #{name}: - You need to add this to TestSetup or use a standard FactoryBot factory."
```
Comment (misaka) [inline spec/support/test_data_cache.rb]: Not critical, but it wouldn't hurt to have the traits here too.

File: spec/support/test_data_cache.rb (hunk 17 above)
Comment (misaka) [reply spec/support/test_data_cache.rb]: Looks good to me.

**Comment by misaka** [comment]
@gpeng

> I still think having to create and recreate large trees of data so often is a smell that too much of the code is coupled to the DB and we're going too heavy on high level tests and not writing code that can be more easily tested in isolation with less reliance on high level tests.

Thing is, I'd rather we rely more on high level tests, and we have a setup that's well understood and common to all the tests. I think that ultimately, the value of a unit test approaches zero with time, whereas the higher level tests generally keep their value for the life of the product.

**Comment by misaka** [comment]
@gpeng 

> Also, I've never seen factory behaviour being tested either.

Ya, I know that that's not common, but I've seen factories not behave in the way their intended before, which is why I suggested we add a few.

File: spec/support/test_data_cache.rb [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+
+  private
+
+    @@cache_populated = false
+    @@courses = nil
+    @@course_factories = {
+        %i[primary unpublished] => -> do
+          FactoryBot.find_or_create(:course, :primary, :unpublished)
+        end,
+    }
```
Comment (misaka) [inline spec/support/test_data_cache.rb]: Move to the top for visibility, they're private as a function of being `@@` variables.

File: spec/support/test_data_cache.rb [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+      @@cache_populated = false
+    end
+
+  private
+
+    @@cache_populated = false
+    @@courses = nil
+    @@course_factories = {
+        %i[primary unpublished] => -> do
+          FactoryBot.find_or_create(:course, :primary, :unpublished)
```
Comment (misaka) [inline spec/support/test_data_cache.rb]: It's probably worth sticking with `create` here, since this factory should always create (i.e. the db should be empty).

File: spec/support/test_data_cache.rb (hunk 19 above)
Comment (TobyRet) [reply spec/support/test_data_cache.rb]: 👍

File: spec/support/test_data_cache.rb (hunk 18 above)
Comment (TobyRet) [reply spec/support/test_data_cache.rb]: 👍

File: spec/support/test_data_cache.rb (hunk 20 above)
Comment (TobyRet) [reply spec/support/test_data_cache.rb]: 👍

File: spec/support/test_data_cache.rb (hunk 21 above)
Comment (TobyRet) [reply spec/support/test_data_cache.rb]: 👍

File: spec/support/test_data_cache.rb [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    def get(name, *traits)
+      case name
+      when :course
+        if @@cache_populated
+          @@courses[traits.sort] || raise_trait_error(:course, traits)
+        else
+          # fall back to factory (while we migrate all our tests)
+          @@course_factories[traits.sort].call
```
Comment (timabell) [inline spec/support/test_data_cache.rb]: What does failure look like if you try and call this with unregistered traits? Would it be easy to figure out what to do to fix it?

Comment (timabell) [reply spec/support/test_data_cache.rb]: I tried it, it's a nill error :cry: 

How about

```suggestion
          if @@course_factories.key?(traits.sort)
              @@course_factories[traits.sort].call
          else
              raise_model_error(name, traits)
          end
```

**Comment by timabell** [comment]
much better, wonder why the CI failed...

File: db/seeds.rb [hunk 23]
```diff
@@ -3,6 +3,8 @@
 
 CourseSubject.destroy_all
 Course.destroy_all
+FinancialIncentive.destroy_all
+SubjectArea.destroy_all
```
Comment (misaka) [inline db/seeds.rb]: Should these be here?

File: app/models/subjects/secondary_subject.rb [hunk 24]
```diff
@@ -20,4 +20,8 @@ class SecondarySubject < Subject
   def self.modern_languages
     @modern_languages ||= find_by(subject_name: "Modern Languages")
   end
+
+  def self.clear_modern_languages_cache
+    @modern_languages = nil
+  end
```
Comment (misaka) [inline app/models/subjects/secondary_subject.rb]: Should this be here?

File: spec/seed_spec.rb [hunk 25]
```diff
@@ -0,0 +1,10 @@
+describe "Seed" do
+  before do
+    SecondarySubject.clear_modern_languages_cache
+  end
+
+  it "seeds without error" do
+    # https://stackoverflow.com/questions/38483820/testing-successful-seed-with-minitest-rspec
+    load Rails.root.join("db/seeds.rb")
+  end
+end
```
Comment (misaka) [inline spec/seed_spec.rb]: Should this be here?

**Comment by timabell** [comment]
squashed in linter fixes
